File: src/ios/CDVBMSClient.swift
/*
    Copyright 2015 IBM Corp.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
        http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

import Foundation
import BMSCore


@objc(CDVBMSClient) class CDVBMSClient : CDVPlugin {

    func initialize(_ command: CDVInvokedUrlCommand) {
#if swift(>=3.0)
        self.commandDelegate!.run(inBackground: {

            guard let region  = command.arguments[0] as? String else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: CustomErrorMessages.invalidRoute)
                // call success callback
                self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
                return
            }

            let client = BMSClient.sharedInstance;

            //use category to handle objective-c exception
            client.initialize(bluemixAppRoute: "", bluemixAppGUID: "", bluemixRegion: region)


            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "")
            // call success callback
            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)

        })
#else
        self.commandDelegate.runInBackground({

            guard let region  = command.arguments[0] as? String else {
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAsString: CustomErrorMessages.invalidRoute)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
                return
            }

            let client = BMSClient.sharedInstance;

            //use category to handle objective-c exception
            client.initialize(bluemixAppRoute: "", bluemixAppGUID: "", bluemixRegion: region)


            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: "")
            // call success callback
            self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
        })
#endif
    }

    func getBluemixAppRoute(_ command: CDVInvokedUrlCommand) {

        #if swift(>=3.0)
            self.commandDelegate!.run(inBackground: {

            let client = BMSClient.sharedInstance
            let backendRoute: String = client.bluemixAppRoute!
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: backendRoute)
            // call success callback
            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
            })
        #else
            self.commandDelegate!.runInBackground({

                let client = BMSClient.sharedInstance
                let backendRoute: String = client.bluemixAppRoute!
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: backendRoute)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif

    }

    func getBluemixAppGUID(_ command: CDVInvokedUrlCommand) {

        #if swift(>=3.0)
        self.commandDelegate!.run(inBackground: {

            let client = BMSClient.sharedInstance
            let backendGUID: String = client.bluemixAppGUID!
            let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: backendGUID)
            // call success callback
            self.commandDelegate!.send(pluginResult, callbackId:command.callbackId)
        })
            #else
            self.commandDelegate!.runInBackground({

                let client = BMSClient.sharedInstance
                let backendGUID: String = client.bluemixAppGUID!
                let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAsString: backendGUID)
                // call success callback
                self.commandDelegate!.sendPluginResult(pluginResult, callbackId:command.callbackId)
            })
        #endif

    }
}

File: src/android/CDVBMSClient.java
/*
    Copyright 2015 IBM Corp.
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
        http://www.apache.org/licenses/LICENSE-2.0
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/
package com.ibm.mobilefirstplatform.clientsdk.cordovaplugins.core;

import android.content.Context;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import com.ibm.mobilefirstplatform.clientsdk.android.core.api.*;
import com.ibm.mobilefirstplatform.clientsdk.android.logger.api.*;

import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

public class CDVBMSClient extends CordovaPlugin {
    private String errorEmptyArg = "Expected non-empty string argument.";
    private static final Logger bmsLogger = Logger.getLogger(Logger.INTERNAL_PREFIX + "CDVBMSClient");

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        boolean ans = true;
        if ("initialize".equals(action)) {
            this.initialize(args, callbackContext);
        } else if ("getBluemixAppRoute".equals(action)) {
            this.getBluemixAppRoute(callbackContext);
        } else if ("getBluemixAppGUID".equals(action)) {
            this.getBluemixAppGUID(callbackContext);
        } else {
            ans = false;
        }
        return ans;
    }


    /**
     * Use the native SDK API to set the base URL for the authorization server.
     *
     * @param args            JSONArray that contains the backendRoute and backendGUID
     * @param callbackContext
     */
    public void initialize(JSONArray args, CallbackContext callbackContext) throws JSONException {
        String region = args.getString(0);
        if (region != null && region.length() > 0) {
            BMSClient.getInstance().initialize(this.cordova.getActivity().getApplicationContext(),region);
            bmsLogger.debug("Successfully initialized BMSClient");
            callbackContext.success();
        } else {
            bmsLogger.error("Trouble initializing BMSClient");
            callbackContext.error(errorEmptyArg);
        }
    }

    private void getBluemixAppRoute(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                String bluemixAppRoute = BMSClient.getInstance().getBluemixAppRoute();
                callbackContext.success(bluemixAppRoute);
            }
        });
    }

    private void getBluemixAppGUID(final CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            public void run() {
                String bluemixAppGUID = BMSClient.getInstance().getBluemixAppGUID();
                callbackContext.success(bluemixAppGUID);
            }
        });
    }
}
